operator/iamkeycontroller/webhook_test.go

```diff
@@ -11,165 +11,150 @@ import (
 		},
 		"GivenIAMKey_WhenWriteConnectionSecretToRefMissing_ThenExpectError": {
-			iamKey: exoscalev1.IAMKey{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: "iamkey-with-buckets",
-				},
-				Spec: exoscalev1.IAMKeySpec{
-					ResourceSpec: xpv1.ResourceSpec{
-						ProviderConfigReference: &xpv1.Reference{
-							Name: "provider-config",
-						},
-						WriteConnectionSecretToReference: nil,
-					},
-					ForProvider: exoscalev1.IAMKeyParameters{
-						Services: exoscalev1.ServicesSpec{
-							SOS: exoscalev1.SOSSpec{
-								Buckets: []string{"bucket.1", "bucket.2"},
-							},
-						},
-					},
-				},
+			connectionSecretToRef: xpv1.SecretReference{},
+			expectedError:         ".spec.writeConnectionSecretToRef.name and .spec.writeConnectionSecretToRef.namespace are required",
+		},
+		"GivenIAMKey_WhenWriteConnectionSecretToRefNameMissing_ThenExpectError": {
+			connectionSecretToRef: xpv1.SecretReference{
+				Namespace: "secret-namespace",
 			},
 			expectedError: ".spec.writeConnectionSecretToRef.name and .spec.writeConnectionSecretToRef.namespace are required",
 		},
-		"GivenIAMKey_WhenWriteConnectionSecretToRefNameMissing_ThenExpectError": {
-			iamKey: exoscalev1.IAMKey{
+		"GivenIAMKey_WhenWriteConnectionSecretToRefNamespaceMissing_ThenExpectError": {
+			connectionSecretToRef: xpv1.SecretReference{
+				Name: "secret-name",
+			},
+			expectedError: ".spec.writeConnectionSecretToRef.name and .spec.writeConnectionSecretToRef.namespace are required",
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			iamKey := exoscalev1.IAMKey{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: "iamkey-with-buckets",
 				},
 				Spec: exoscalev1.IAMKeySpec{
 					ResourceSpec: xpv1.ResourceSpec{
-						ProviderConfigReference: &xpv1.Reference{
-							Name: "provider-config",
-						},
-						WriteConnectionSecretToReference: &xpv1.SecretReference{
-							Name:      "",
-							Namespace: "secret-namespace",
-						},
-					},
-					ForProvider: exoscalev1.IAMKeyParameters{
-						Services: exoscalev1.ServicesSpec{
-							SOS: exoscalev1.SOSSpec{
-								Buckets: []string{"bucket.1", "bucket.2"},
-							},
-						},
+						ProviderConfigReference: &xpv1.Reference{Name: "provider-config"},
+						// connection secret to ref is being tested
+						WriteConnectionSecretToReference: &tc.connectionSecretToRef,
 					},
+					ForProvider: exoscalev1.IAMKeyParameters{Services: exoscalev1.ServicesSpec{SOS: exoscalev1.SOSSpec{Buckets: []string{"bucket.1"}}}},
 				},
+			}
+			validator := &IAMKeyValidator{log: logr.Discard()}
+			err := validator.ValidateCreate(nil, &iamKey)
+			if tc.expectedError != "" {
+				assert.EqualError(t, err, tc.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestIAMKeyValidator_ValidateCreate_RequireProviderConfigToRef(t *testing.T) {
+	tests := map[string]struct {
+		iamKey              exoscalev1.IAMKey
+		providerConfigToRef xpv1.Reference
```

Comment:
Also here, please add a test case to check that the `nil` is handled correctly as well.
```suggestion
		providerConfigToRef *xpv1.Reference
```